File: Code_Examples/Chapter_02/Src/C/area.acc.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

# define NPOINTS 4000
# define MAXITER 4000


struct complex{
  double real;
  double imag;
};

int main(){
  int i, j, iter, numoutside = 0, found;
  double area, error, ztemp;
  struct complex z, c;

  clock_t start, end;
  double cpu_time_used;
     
  start = clock();
/*
 *   
 *
 *     Outer loops run over npoints, initialise z=c
 *
 *     Inner loop has the iteration z=z*z+c, and threshold test
 */
#pragma acc parallel loop reduction(+:numoutside) private(i,j) 
  for (i=0; i<NPOINTS; i++) {
#pragma acc loop vector private(c,z,found,c,z,iter,ztemp)
    for (j=0; j<NPOINTS; j++) {
      c.real = -2.0+2.5*(double)(i)/(double)(NPOINTS)+1.0e-7;
      c.imag = 1.125*(double)(j)/(double)(NPOINTS)+1.0e-7;
      z.real=c.real;
      z.imag=c.imag;
      found  = 0;
      iter = 0;
      while(iter<MAXITER && found == 0){
	ztemp=(z.real*z.real)-(z.imag*z.imag)+c.real;
	z.imag=z.real*z.imag*2+c.imag; 
	z.real=ztemp; 
	if ((z.real*z.real+z.imag*z.imag)>4.0e0 && found == 0) {
	  numoutside++; 
          found = 1;
	}
	iter =  iter + 1;
      }
    }
  }



  end = clock();
  cpu_time_used = ((double) (end - start)) / CLOCKS_PER_SEC;



/*
 *  Calculate area and error and output the results
 */

   area=2.0*2.5*1.125*(double)(NPOINTS*NPOINTS-numoutside)/(double)(NPOINTS*NPOINTS);
   error=area/(double)NPOINTS;

   printf("Area of Mandlebrot set = %12.8f +/- %12.8f\n",area,error);
   printf("Time taken for calculation: %f\n",cpu_time_used);    
}

File: Code_Examples/Chapter_02/Src/Fortran/area.acc.f90
program area_mandelbrot
  implicit none
!
  integer, parameter :: sp = kind(1.0)   
  integer, parameter :: dp = kind(1.0d0)
  integer :: i, j, iter, numoutside
  logical :: found
  integer, parameter :: npoints = 4000, maxiter = 4000
  real (kind=dp) :: area, error
  complex (kind=dp) :: c , z
  real :: starttime, endtime
!
! Calculate area of mandelbrot set
!
!    Outer loops runs over npoints, initialize z=c
!
!    Inner loop has the iteration z=z*z+c, and threshold test
!

  numoutside = 0 

  call cpu_time(starttime)
  
  !$acc parallel loop gang reduction(+:numoutside) private(c,z,iter,i,j,found)
  do i = 0,npoints-1
     !$acc loop vector
     do j= 0,npoints-1 
        c = cmplx(-2.0+(2.5*i)/npoints + 1.0d-07,(1.125*j)/npoints + 1.0d-07)
        z = c
        iter = 0
        found = .false.
        do while (iter < maxiter .and. .not. found) 
           z = z*z + c 
           iter = iter + 1
           if (real(z)*real(z)+aimag(z)*aimag(z) > 4) then
              numoutside = numoutside + 1 
              found = .true.
           endif
        end do 
     end do
  end do


  call cpu_time(endtime)
!
! Output results
!
  area = 2.0*2.5*1.125 * real(npoints*npoints-numoutside)/real(npoints*npoints)
  error = area/real(npoints)
  print *, "Area of Mandelbrot set = ",area," +/- ",error
  print *, "Time taken for calculation: ", endtime - starttime
  !
  stop
end program area_mandelbrot
